src/comm.jl

```diff
@@ -1,264 +1,1887 @@
    # assume communication to 6 neighbors by default
 using MPI
 function  get_neighbors(domain::Domain)
-   rowMin = rowMax = colMin = colMax = planeMin = planeMax = true
-
-   domain.m_rowLoc == 0 ? rowMin = false : rowMin = true
-   domain.m_rowLoc == domain.m_tp - 1 ? rowMax = false : rowMax = true
-   domain.m_colLoc == 0 ? colMin = false : colMin = true
-   domain.m_colLoc == domain.m_tp - 1 ? colMax = false : colMax = true
-   domain.m_planeLoc == 0 ? planeMin = false : planeMin = true
-   domain.m_planeLoc == domain.m_tp - 1 ? planeMax = false : planeMax = true
+   rowMin = !(domain.m_rowLoc == 0)
+   rowMax = !(domain.m_rowLoc == domain.m_tp - 1)
+   colMin = !(domain.m_colLoc == 0)
+   colMax = !(domain.m_colLoc == domain.m_tp - 1)
+   planeMin = !(domain.m_planeLoc == 0)
+   planeMax = !(domain.m_planeLoc == domain.m_tp - 1)
 
    return rowMin, rowMax, colMin, colMax, planeMin, planeMax
 end
 
-# doRecv flag only works with regular block structure
-function commRecv(
-      domain::Domain, msgType::Int, xferFields::IndexT,
-      dx::IndexT, dy::IndexT, dz::IndexT, doRecv::Bool, planeOnly::Bool, comm::MPI.Comm
-   )
+# Assume 128 byte coherence
+# Assume Float64 is an "integral power of 2" bytes wide
+const CACHE_COHERENCE_PAD_REAL = div(128, sizeof(Float64))
+
+function commRecv(domain::Domain, msgType, xferFields, dx, dy, dz, doRecv, planeOnly)
+   if domain.comm === nothing
+       return
+   end
+
+   # post recieve buffers for all incoming messages
    maxPlaneComm = xferFields * domain.maxPlaneSize
    maxEdgeComm  = xferFields * domain.maxEdgeSize
+   pmsg = 0 # plane comm msg
+   emsg = 0 # edge comm msg
+   cmsg = 0 # corner comm msg
+
+   baseType = MPI.Datatype(Float64) # TODO support Float32
+
+   # assume communication to 6 neighbors by default
    rowMin, rowMax, colMin, colMax, planeMin, planeMax = get_neighbors(domain)
-   pmsg = emsg = cmsg = 1
-   # for r in domain.recvRequest
-   #    r = MPI.REQUEST_NULL
-   # end
-   myRank = getMyRank(comm)
+
+   # XXX: empty recvRequests?
+   # for (Index_t i=0; i<26; ++i) {
+   #   domain->recvRequest[i] = MPI_REQUEST_NULL ;
+   # }
+
+   myRank = MPI.Comm_rank(domain.comm)
 
    # post receives
+   function irecv!(fromProc, offset, recvCount)
+      idx = offset + 1
+      data = view(domain.commDataRecv, idx:(idx+recvCount))
+      return MPI.Irecv!(data, fromProc, msgType + fromProc, MPI.MPI_COMM_WORLD)
```

Comment:
Adding fromproc is different: https://github.com/wsmoses/Enzyme-MPI/blob/d211d3522628628cc3990a2470639b3159104a37/lulesh-comm.cc#L112

Reply:
Seeing them side by side I withdraw my objection. The transfer is already tagged from which rank it is coming and thus does not need `+fromProc` mixing.